README.md
# supSMSTAS
<p align="center">
  <span>English</span> |
  <a href="https://github.com/sup39/supSMSTAS/blob/main/README.ja.md">日本語</a>
</p>

*A tool to **sup**port **S**uper **M**ario **S**unshine **T**ool **A**ssisted **S**peedrun/Superplay*

## Installation/Update
```
pip install -U supSMSTAS
```
:warning: Python 3.8 or above is required

## Usage
```
python -m supSMSTAS
```

## Required Gecko code
You need to add and activate the following Gecko code in Dolphin:
```
C20ECDE0 00000004
3C60817F 80631000
7C631B79 38600258
41820008 38600096
60000000 00000000
C20ED284 00000005
4E800021 3D60817F
816B1010 55608380
2800817E 4082000C
7D6803A6 4E800021
60000000 00000000
```

Make sure to disable those codes that
alter QF sync (advance by QF) or render triangle/hitbox
if you are using any of them.

If the checkboxes in the Runtime tab don't work,
try to disable all Gecko codes in Dolphin (except the *required Gecko code* mentioned above).

## Execute the source code directly
To install dependencies, the easiest way is to run the following command
in the root directory of the package:
```
pip install .
```

Due to some module/script problems,
in order to start the program,
instead of using `python path/to/supSMSTAS/__main__.py` (as script),
make sure to use `python -m path.to.supSMSTAS` (as module) outside the `src/supSMSTAS/` directory.
For example, if you are in the `src/` directory, you can run the following command:
```
python -m supSMSTAS
```


CHANGELOG.md
# CHANGELOG
## \[v0.0.1.2] consider the -1 area of triangles (2022/07/20)
- WFC
  - consider the -1 area of triangles to improve the precision of WFC plot
- SMSDolphin
  - add `*args` and `**kwargs` to `hook()`
  - use sup-dolphin-memory-lib >= 0.1.2 to solve pid problem
## \[v0.0.1.1] improve performance by using PolyCollection (2022/07/19)
- WFC
  - improve process speed of `make_geo_plot()` by using PolyCollection
  - render dynamic triangle hitbox
- README
  - add Japanese version of README
## \[v0.0.1] init (2022/06/26)
- WFC
  - plot WFC hitbox (x-z plot and x/z-y plot) with matplotlib toolbar
  - options:
    - track Mario
    - show Mario as red point
    - airborne or grounded
    - on Yoshi
    - invert X/Z
  - angle dial for x/z-y plot
- Runtime
  - advance by QF
  - show Hitbox


src/supSMSTAS/WFC.py
# SPDX-License-Identifier: GPL-3.0-only
# Copyright (c) 2022 sup39

import numpy as np
from numpy import array
from .shape import Polyhedron
from matplotlib.collections import PolyCollection

class Surface:
  def __init__(self, surtype, surpara, trntype, unk7, verts, vidxs=None, n=None, c=None):
    self.surtype, self.surpara, self.trntype, self.unk7 = \
      surtype, surpara, trntype, unk7
    self.vidxs = vidxs
    self.verts = verts
    self.minY = verts[:,1].min()
    self.maxY = verts[:,1].max()
    self.n = normalize(np.cross(verts[1]-verts[0], verts[2]-verts[1])) if n is None else n
    self.c = -np.dot(verts[0], self.n) if c is None else c
  def __repr__(self):
    return 'minY=%.0f maxY=%.0f n=(%5.2f, %5.2f, %5.2f)'%(
      self.minY, self.maxY, *self.n,
    )

makeTriPrism = lambda tri0, tri1: Polyhedron([
  *tri0,
  *tri1,
], [
  (0, 1), (1, 2), (2, 0),
  (0, 3), (1, 4), (2, 5),
  (3, 4), (4, 5), (5, 3),
])

def extendTriangle(tri, axis):
  ans = np.array(tri, 'd')
  tri = tri[:, axis]
  for i in [-1, 0, 1]:
    l1, l2 = tri[i]-tri[i+1], tri[i]-tri[i-1]
    ans[i, axis] += (l1+l2)/np.abs(np.cross(l1, l2).item())
  return ans

def makeGround(tri, hG=0):
  vertsE = extendTriangle(tri.verts, [0, 2])
  vertsLow = vertsE-(0,108,0)
  ySlice = tri.minY-30
  v = array(vertsLow)
  vertsLow[:,1] = np.clip(vertsLow[:,1], a_min=ySlice, a_max=None)
  verts = [
    *vertsLow,
    *(vertsE+(0,hG,0)),
  ]
  edges = [
    # top face
    (3, 4), (4, 5), (5, 3),
    # vertical edge
    (0, 3), (1, 4), (2, 5),
  ]
  r = v[:, 1]-ySlice
  b = r>0
  bc = b.sum()
  if bc in [0, 3]:
    # bottom face
    edges += [(0, 1), (1, 2), (2, 0)]
  else:
    k0 = np.where(b if bc==1 else ~b)[0][0]
    k1 = (k0+1)%3
    k2 = (k0+2)%3
    v0, r0 = v[k0], r[k0]
    v1, r1 = v[k1], r[k1]
    v2, r2 = v[k2], r[k2]
    verts += [
      (r1*v0-r0*v1)/(-r0+r1),
      (r2*v0-r0*v2)/(-r0+r2),
    ]
    edges += [
      (k0, 6), (6, k1),
      (k0, 7), (7, k2),
      (k1, k2), (6, 7),
    ]
  poly = Polyhedron(verts, edges)
  return poly
def makeRoof(tri, hR=82):
  verts = extendTriangle(tri.verts, [0, 2])
  poly = makeTriPrism(verts-(0,hR,0), verts-(0,160,0))
  return poly
def makeWall(tri, rW=50, dy=30):
  n = tri.n
  isXWall = np.abs(n[0])>0.707
  verts = extendTriangle(tri.verts, [2 if isXWall else 0, 1]) - (0, dy, 0)
  off = (np.abs(rW/n[0]),0,0) if isXWall else (0,0,np.abs(rW/n[2]))
  poly = makeTriPrism(verts-off, verts+off)
  return poly

def make_geo_plot(ax, hitboxs, p0, pn, axes):
  # paras
  arrowWidthBase = 70
  arrowWidthMul = 0.5
  arrowLenMax = 200 #80
  arrowLenTher = 200
  arrowLenMul1 = 0.7
  arrowLenMul2 = 0.35
  arrowHeadLenMul = 0.3
  arrowLenMul2off = 0.1 #0.025
  arrowCountMax = 100
  aw, ah = 0.2, 0.6 # width, height of bottom rect
  arrowVerts = array([
    [0.0,  aw],
    [ ah,  aw],
    [ ah, 0.5],
    [1.0, 0.0],
    [ ah, -.5],
    [ ah, -aw],
    [0.0, -aw],
  ])

  # plot
  patches = []
  for polys, awmul, alen, facecolor, arcolor in hitboxs:
    # draw wall hitboxs (draw in reverse order)
    for poly, n in polys[::-1]:
      # clip at y=yy and take (x, z) coordinate
      verts = poly.slicePlane(p0, pn)[:,axes]
      if len(verts) == 0: continue
      # plot hitbox area
      patches.append((verts, facecolor, 'black')) # (verts, fc, ec)
      # plot arrow
      n = n*(1, 0, 1) # no y arrow
      n = n[axes]
      nnorm = np.linalg.norm(n)
      if nnorm == 0: continue # TODO
      n = n/nnorm
      l = array([n[1], -n[0]])
      ## basis matrix
      B = np.vstack([n, l])
      ## center
      C = verts.mean(axis=0)
      vertsO = verts-C
      ## n-l coordinates
      nn = np.dot(vertsO, n)
      ll = np.dot(vertsO, l)
      ## determine arrow width and count
      iminl = ll.argmin()
      imaxl = ll.argmax()
      minl = ll[iminl]
      maxl = ll[imaxl]
      dl = maxl-minl
      arrowWidth = arrowWidthBase*awmul
      nArrow = np.floor(dl/arrowWidth).astype('i')
      if nArrow >= 1:
        if nArrow>arrowCountMax: nArrow = arrowCountMax
      else:
        arrowWidth = dl
        nArrow = 1
      arrowWidth *= arrowWidthMul
      ## determine arrow position
      vlen = len(verts)
      nbss = []
      ls = np.linspace(minl, maxl, nArrow+2)[1:-1]
      for step in [1, -1]:
        i0 = iminl
        i1 = (i0+step)%vlen
        nbs = []
        for lb in ls:
          while ll[i1] < lb: i0, i1 = i1, (i1+step)%vlen
          r0 = lb-ll[i0]
          r1 = ll[i1]-lb
          nbs.append((nn[i0]+nn[i1])/2 if r0+r1==0 else (nn[i0]*r1+nn[i1]*r0)/(r0+r1))
        nbss.append(nbs)
      nbss = array(nbss)
      ncts = nbss.mean(axis=0)
      nrgs = np.abs(np.diff(nbss, axis=0)[0])
      # add arrow
      maty = array([-n[1], n[0]])*arrowWidth
      for p, nrg in zip(
        C+np.matmul(np.column_stack([ncts, ls]), B),
        nrgs,
      ):
        if nrg > arrowLenTher:
          nrg = min(arrowLenMax, nrg*arrowLenMul2)
          offs = [-arrowLenMul2off, 1+arrowLenMul2off]
        else:
          nrg = min(arrowLenMax, nrg*arrowLenMul1)
          offs = [0.5]
        dn = n * nrg
        # add patches
        verts0 = np.matmul(arrowVerts, [n*nrg, maty])
        for off in offs:
          patches.append((
            p-dn*off+verts0,
            arcolor, # fc
            arcolor, # ec
          ))

  if len(patches):
    ax.add_collection(PolyCollection(
      **{
        k: v
        for k, v in zip(('verts', 'facecolors', 'edgecolors'), zip(*patches))
      },
    ))

  ax.set_xlabel('xyz'[axes[0]])
  ax.set_ylabel('xyz'[axes[1]])

  return axes


src/supSMSTAS/shape.py
# SPDX-License-Identifier: MIT
# Copyright (c) 2022 sup39[サポミク]

import numpy as np
array = np.array
import matplotlib.pyplot as plt
from matplotlib.path import Path
from matplotlib import patches
def normalize(x):
  l = np.linalg.norm(x)
  return x if l==0 else x/l

# 多角形
class Polygon():
  def __init__(self, verts):
    '''
    * self.verts:
      頂点の配列
      Array of vertices
    '''
    self.verts = np.array(verts)
  def clipLine(self, p, n, c):
    '''
    半平面 (x-p)・n >= c との共通部分を取る
    Take intersection with half plane (x-p)・n >= c
    * p:
      直線上の一点
      Any point on the line
    * n:
      直線の法線ベクトル
      Normal vector of the line
    * c:
      定数
      Constant
    '''
    p, n, c = map(array, (p, n, c))
    r = np.dot(self.verts-p, n)-c
    verts = []
    for i in range(len(r)):
      v0, r0 = self.verts[i-1], r[i-1]
      v1, r1 = self.verts[i], r[i]
      if r1 >= 0:
        if r0 < 0: # (- +)
          verts.append((r1*v0-r0*v1)/(-r0+r1))
        verts.append(v1)
      elif r0 >= 0: # (+ -)
        verts.append((-r1*v0+r0*v1)/(r0-r1))
    self.verts = np.array(verts) if len(verts) else self.verts[:0]
  @property # getter
  def path(self):
    if self.verts.shape[0] == 0: return None
    verts = self.verts
    assert verts.shape[-1] == 2, 'verts should be 2D'
    return Path([*verts, (0, 0)], [
      Path.MOVETO,
      *(Path.LINETO for _ in range(1, verts.shape[0])),
      Path.CLOSEPOLY,
    ])
  def plot(self, margin=0.05, facecolor='#2ee5b8', lw=1):
    '''
    この多角形を描画し、figとaxを返す
    Plot this polygon and return ``fig'' and ``ax''
    * margin:
      Set margin of the plot
    * facecolor:
      面の色
      Face color
    * lw:
      線の太さ
      Line width
    '''
    fig, ax = plt.subplots()
    if self.verts.shape[0] == 0: return ax
    # path
    path = self.path
    patch = patches.PathPatch(path, facecolor=facecolor, lw=lw)
    ax.add_patch(patch)
    verts = self.verts
    xMax, yMax = verts.max(axis=0)
    xMin, yMin = verts.min(axis=0)
    xMg, yMg = verts.ptp(axis=0)*margin
    ax.set_xlim(xMin-xMg, xMax+xMg)
    ax.set_ylim(yMin-yMg, yMax+yMg)
    return fig, ax
  def __repr__(self):
    return 'Polygon with %d vertices:\n%s'%(
      len(self.verts),
      array(self.verts, 'f'),
    )

# 多面体
class Polyhedron:
  def __init__(self, verts, edges):
    '''
    * self.verts:
      頂点の配列
      Array of vertices
    * self.edges:
      辺(2頂点の番号)の配列
      Array of edges (indices of 2 vertices)
    '''
    self.verts = np.array(verts)
    self.edges = np.array(edges)
  def clipPlane(self, p, n, c=0):
    '''
    半空間 (x-p)・n >= c との共通部分を取る
    Take intersection with half space (x-p)・n >= c
    * p:
      平面上の一点
      Any point on the plane
    * n:
      直線の法線ベクトル
      Normal vector of the plane
    * c:
      定数
      Constant
    '''
    p = array(p)
    n = array(n)
    r = np.dot(self.verts-p, n)-c
    rb = [s>=0 for s in r]
    # map vertex indices old to new
    io2n = {
      iO: iN
      for iN, iO in enumerate(iO for iO, sb in enumerate(rb) if sb)
    }
    # handle old vert
    verts = [v for v, sb in zip(self.verts, rb) if sb]
    edges = []
    for i0, i1 in self.edges:
      if rb[i0] and rb[i1]:
        # remain
        edges.append((io2n[i0], io2n[i1]))
      elif rb[i0] or rb[i1]:
        # new vert
        v0, r0 = self.verts[i0], abs(r[i0])
        v1, r1 = self.verts[i1], abs(r[i1])
        vN = (r1*v0+r0*v1)/(r0+r1)
        edges.append((io2n[i0 if rb[i0] else i1], len(verts)))
        verts.append(vN)
      # else drop edge
    # add new face
    nOld = len(io2n)
    vNews = verts[nOld:]
    if len(vNews):
      assert len(vNews) >= 3
      p0, p1 = vNews[:2]
      # choose p1-p0 as e1
      e1 = normalize(p1-p0)
      # choose e2 that ⊥ n, e1
      e2 = normalize(np.cross(n, e1))
      # set (p0+p1)/2 as new origin, and use {e1, e2} as new basis
      cNews = np.dot(vNews-(p0+p1)/2, array([e1, e2]).transpose())
      # indices of new verts CCW
      jNews = nOld+np.arctan2(cNews[:,0], cNews[:,1]).argsort()
      # add to edge
      for i in range(len(vNews)):
        edges.append((jNews[i-1], jNews[i]))
    # final
    self.verts = array(verts)
    self.edges = array(edges)
  def slicePlane(self, p, n):
    '''
    平面 (x-p)･n=0 との共通部分(多角形)の頂点を返す
    Return vertices of intersection(polygon) with plane (x-p)･n=0
    * p:
      平面上の一点
      Any point on the plane
    * n:
      平面の法線ベクトル
      Normal vector of the plane
    '''
    p = array(p)
    n = array(n)
    r = np.dot(self.verts-p, n)
    # handle old verts
    #for i0, i1 in self.edges:
    #  # two vertices on other side of the plane
    #  if np.sign(r[i0]) != np.sign(r[i1]):
    #    v0, r0 = self.verts[i0], abs(r[i0])
    #    v1, r1 = self.verts[i1], abs(r[i1])
    #    vN = (r1*v0+r0*v1)/(r0+r1)
    #    vNews.append(vN)
    signr = np.sign(r[self.edges])
    edgesNew = self.edges[signr[:,0]!=signr[:,1]]
    vv = self.verts[edgesNew]
    rr = np.abs(r[edgesNew])
    vNews = ((vv[:,0]*rr[:,1,None])+(vv[:,1]*rr[:,0,None]))/rr.sum(axis=1)[:,None]
    # new verts
    if len(vNews):
      assert len(vNews) >= 2
      p0, p1 = vNews[:2]
      e1 = normalize(p1-p0)
      e2 = normalize(np.cross(n, e1))
      cNews = np.dot(vNews-(p0+p1)/2, array([e1, e2]).transpose())
      jNews = np.arctan2(cNews[:,0], cNews[:,1]).argsort()
      return vNews[jNews]
    else:
      return vNews
  def __repr__(self):
    return 'Polyhedron with %d vertices and %d edges:\n%s'%(
      len(self.verts), len(self.edges),
      array(self.verts[self.edges], 'f')
    )


src/supSMSTAS/UI.py
# SPDX-License-Identifier: GPL-3.0-only
# Copyright (c) 2022 sup39

import os
import time
## numpy
import numpy as np
from numpy import array
## matplotlib
import matplotlib as mpl
mpl.use('Qt5Agg')
from matplotlib.backends.backend_qt5agg import FigureCanvasQTAgg, NavigationToolbar2QT
from matplotlib.figure import Figure
import matplotlib.patches as patches
## PyQt5
from PyQt5.QtCore import *
from PyQt5.QtWidgets import *
## logging
import logging
## local
from .WFC import *

# add TRACE
logging.TRACE = 5
logging.addLevelName(logging.TRACE, 'TRACE')
def trace(self, message, *args, **kws):
  if self.isEnabledFor(logging.TRACE):
    self._log(logging.TRACE, message, args, **kws)
logging.Logger.trace = trace
# logger
logger = logging.getLogger('supSMSTAS')

# widgets utils
def layoutAddItem(ll, w):
  if w is None: ll.addStretch()
  elif isinstance(w, QLayout): ll.addLayout(w)
  else: ll.addWidget(w)
def HBox(*children, **kwargs):
  ll = QHBoxLayout(**kwargs)
  for w in children: layoutAddItem(ll, w)
  return ll
def VBox(*children, **kwargs):
  ll = QVBoxLayout(**kwargs)
  for w in children: layoutAddItem(ll, w)
  return ll

class MPLCanvas(FigureCanvasQTAgg):
  def __init__(self, parent=None, width=5, height=5, dpi=100):
    fig = self.fig = Figure(figsize=(width, height), dpi=dpi)
    self.axs = [
      fig.add_subplot(121),
      fig.add_subplot(122),
    ]
    super().__init__(fig)

# Dolphin for SMS
from dolphin.memorylib import Dolphin
class SMSDolphin(Dolphin):
  symbols = { # TODO
    b'GMSJ01\x00\x00': {
      'gpMarioOriginal': 0x8040A378,
      'gpMap': 0x8040A570,
    }
  }
  def get_symb_addr(self, name): # override
    verID = self.memory.buf[:8].tobytes()
    return SMSDolphin.symbols[verID][name]

  def __init__(self):
    super().__init__()
    self.hooked = False
  def hook(self, *args, **kwargs):
    self.hooked = False
    if super().hook(*args, **kwargs) is None:
      return 'SMS is not running'
    # found pid -> check game
    verID = self.memory.buf[:8].tobytes()
    if verID[:3] != b'GMS':
      return 'Current game is not SMS'
    if verID not in SMSDolphin.symbols:
      return 'Only NTSJ-1.0 is supported in this version'
    self.hooked = True

  def checkList2list(self, ptr):
    ans = []
    while ptr:
      ptr, data = self.read_struct(ptr+4, '>II')
      ans.append(Surface(
        *self.read_struct(data, '>HHBB'),
        array(self.read_struct(data+0x10, '>9f'), 'f').reshape(3, 3),
        n=array(self.read_struct(data+0x34, '>3f'), 'f'),
        c=self.read_float(data+0x40),
      ))
    return ans

# widgets
class WFCWidget(QWidget):
  def __init__(self, dolphin, parent=None):
    super().__init__(parent=parent)
    self.d = dolphin
    self._init_state()
    self._init_layout()
  def activate(self, val):
    # stop timer if deactivate
    self.setFPS(self.fps) if val else self.timer.stop()
  def _init_state(self):
    self.trackMario = Qt.Checked
    self.showMario = Qt.Checked
    self.airborne = Qt.Checked
    self.yoshi = 0
    self.fps = 8
    self.updating = False
    self.invertX = 0
    self.invertZ = 0
    self.xzAngle = 0
  def _init_layout(self):
    # timer
    timer = self.timer = QTimer()
    timer.timeout.connect(self.updatePlot)
    # mcv
    self.mcv = mcv = MPLCanvas(self, width=11, height=5, dpi=100)
    self.updatePlot()
    # toolbar
    toolbar = NavigationToolbar2QT(mcv, self)
    # buttons
    llbtn = QHBoxLayout()
    btn = QPushButton()
    btn.setText('Update plot')
    btn.clicked.connect(self.updatePlot)
    llbtn.addWidget(btn)
    # update
    lbFPS = QLabel()
    lbFPS.setText('fps')
    sbFPS = QSpinBox()
    sbFPS.setRange(0, 60)
    sbFPS.setValue(self.fps)
    #cbFPS.setText('fps')
    sbFPS.valueChanged.connect(lambda: self.setFPS(sbFPS.value()))
    self.setFPS(self.fps)
    ## add widget
    llbtn.addWidget(lbFPS)
    llbtn.addWidget(sbFPS)
    for text, attr in [
      ('Track Mario', 'trackMario'),
      ('Show Mario', 'showMario'),
      ('Airborne', 'airborne'),
      ('Yoshi', 'yoshi'),
      ('invert X', 'invertX'),
      ('invert Z', 'invertZ'),
    ]:
      cb = QCheckBox()
      cb.setText(text)
      cb.setCheckState(getattr(self, attr))
      cb.clicked.connect((lambda attr, cb: lambda: setattr(self, attr, cb.checkState()))(attr, cb))
      llbtn.addWidget(cb)
    llbtn.addStretch()
    # angle
    lbXZAngle = QLabel()
    lbXZAngle.setText('Angle')
    dlXZAngle = QDial()
    dlXZAngle.setWrapping(True)
    #dlXZAngle.setNotchesVisible(True) # TODO
    dlXZAngle.valueChanged.connect(lambda: self.setXZAngle(dlXZAngle.value()))
    ## angle button
    llXZAngleBtns = QVBoxLayout()
    for text, val in [
      ('X+', 75),
      ('X-', 25),
      ('Z+', 50),
      ('Z-', 0),
    ]:
      btn = QPushButton()
      btn.setText(text)
      btn.clicked.connect((lambda val: lambda: dlXZAngle.setValue(val))(val))
      llXZAngleBtns.addWidget(btn)
    # footer
    llFooter = HBox(
      # buttons
      VBox(
        llbtn,
        None,
      ),
      None,
      ## angle
      lbXZAngle,
      dlXZAngle,
      llXZAngleBtns,
    )
    # layout
    self.setLayout(VBox(
      toolbar,
      mcv,
      llFooter,
    ))
  def toggleTimer(self, toggle):
    if toggle:
      self.timer.start()
    else:
      self.timer.stop()
  def setFPS(self, fps):
    self.fps = fps
    if fps == 0:
      self.timer.stop()
    else:
      self.timer.setInterval(int(1000/fps))
      self.timer.start()
  def setXZAngle(self, val):
    self.xzAngle = val
    #self.updatePlot()
  def updatePlot(self):
    if self.updating or not self.d.hooked: return
    self.updating = True
    try: self._updatePlot()
    except:
      import traceback
      traceback.print_exc()
    self.updating = False
  def _updatePlot(self):
    d = self.d
    pos = d.read_struct(('gpMarioOriginal', 0x10), '>3f')
    if pos is None: return
    pos = array(pos)
    x, y, z = pos
    # get collision data (static collision)
    colInfo = d.read_struct(('gpMap', 0x10, 0), '>ffI4x4xII')
    if colInfo is None: return
    xLimit, zLimit, xBlockCount, ptrStCLR, ptrDyCLR = colInfo
    ## TBGCheckListRoot[zBlockCount][xBlockCount]
    colOff = int((z+zLimit)//1024*xBlockCount + (x+xLimit)//1024)*36
    ## root->ground(12*2).next(4)

    # mario
    yoshi = self.yoshi
    airborne = self.airborne
    rW = 80 if yoshi else 50

    t0 = time.time()
    stGnds, stRoofs, stWalls = [
      [
        tri
        for ptrCLR in [ptrStCLR, ptrDyCLR]
        for tri in d.checkList2list(d.read_uint32(ptrCLR+colOff+4+12*j))
      ]
      for j in range(3)
    ]
    hitboxs = [
      ([
        #(c, array((0, -1, 0)))
        (makeRoof(tri, 82 if airborne else 2), array([0, -1, 0]))
        for tri in stRoofs
        #for c in makeCol(tri, airborne=airborne, yoshi=yoshi)
      # if tri.maxY >= yMin
      ##], 1.0, 78.0 if airborne else 158.0, '#f88e', '#800'),
      ## FIXME: draw arrow
      ], 1.0, 0, '#f88e', '#800'),
      ([
        #(c, array((0, 1, 0)))
        (makeGround(tri, 0 if airborne else 0), array([0, 1, 0])) # TODO hG=100
        for tri in stGnds # TODO grounded hitbox
        #for c in makeCol(tri, airborne=airborne, yoshi=yoshi)
      # if tri.maxY >= yMin
      ## FIXME: draw arrow
      ], 1.0, 0, '#88fe', '#008'), # TODO
      *(
        ([
          #(c, tri.n)
          (makeWall(tri, w, dy), tri.n)
          for tri in stWalls
          #for c in [makeCol(tri, airborne=airborne, yoshi=yoshi)[ii]]
        # if tri.maxY >= yMin
        # arrow size scale (0.5 if radius is 25)
        ], 1.0 if airborne or ii == 1 else 0.5, rW, ['#8fcc', '#8f8c'][ii], ['#084', '#080'][ii])
        #], 1.0 if airborne or ii == 1 else 0.5, rW, '#8f8e', '#080')
        for ii, (w, dy) in zip((1, 0), [(rW, 30), (rW, 150)] if airborne else [(rW, 60), (rW/2, 30)])
      ),
    ]
    # xzAngle
    theta = self.xzAngle/50*np.pi
    pnXZ = (-np.sin(theta), 0, -np.cos(theta))
    axesXZ = [2 if abs(pnXZ[0])>abs(pnXZ[2]) else 0, 1]
    invertX = self.invertX==Qt.Checked
    invertZ = self.invertZ==Qt.Checked
    # ax
    xzAngle = self.xzAngle
    for ax, (pn, axes) in zip(self.mcv.axs, [
      ((0, 1, 0), [0, 2]),
      (pnXZ, axesXZ),
    ]):
      ax.patches.clear()
      ax.collections.clear()
      make_geo_plot(ax, hitboxs, pos, pn, axes)
      if self.showMario:
        ax.add_patch(patches.Circle(pos[axes], 25, fc='red'))
      if self.trackMario:
        offs = [(-1000, 1000), (-1000, 1000), (-1000, 1000)] # TODO
        for f, i in zip((ax.set_xlim, ax.set_ylim), axes):
          p = pos[i]
          d0, d1 = offs[i]
          if invertX & (i==0) | invertZ & (i==2): # invert
            d0, d1 = d1, d0
          f(p+d0, p+d1)
      ax.grid(True)
    # apply
    self.mcv.fig.tight_layout()
    self.mcv.draw()
    logger.trace('%.2f'%((time.time()-t0)*1000))

class RuntimeWidget(QWidget):
  def __init__(self, dolphin, parent=None):
    super().__init__(parent=parent)
    self.d = dolphin
    self._init_state()
    self._init_layout()
  def _init_state(self):
    pass
  def _init_layout(self):
    # QF
    cbQF = QCheckBox()
    cbQF.setText('Advance by QF')
    cbQF.clicked.connect(lambda: self.toggleQFSync(cbQF.checkState()))
    # hitbox
    cbHitbox = QCheckBox()
    cbHitbox.setText('Show hitbox')
    cbHitbox.clicked.connect(lambda: self.toggleHitbox(cbHitbox.checkState()))
    # layout
    self.setLayout(VBox(
      cbQF,
      cbHitbox,
      None,
    ))
  def toggleQFSync(self, val):
    self.d.write_uint8(0x817F1000, 1 if val else 0)
  def toggleHitbox(self, val):
    ptrAddr = 0x817F1010
    funcAddr = 0x817E05B8 # TODO
    baseAddr = 0x817E0000
    if val: # TODO other scripts
      baseDir = os.path.dirname(__file__)
      with open(os.path.join(baseDir, 'codes/showHitbox.bin'), 'rb') as fr:
        buf = fr.read()
      self.d.memory.buf[baseAddr-0x80000000:baseAddr-0x80000000+len(buf)] = buf
      self.d.write_uint32(ptrAddr, funcAddr)
    else:
      self.d.write_uint32(ptrAddr, 0)

class MainWindow(QMainWindow):
  def __init__(self, dolphin=None, *args, **kwargs):
    super().__init__(*args, **kwargs)
    self.setWindowTitle("supSMSTAS")
    self.dolphin = SMSDolphin() if dolphin is None else dolphin
    self._init_layout()
    self.hook()
  def _init_layout(self):
    # hook button bar: btHook, lbHook, (spacer)
    btHook = QPushButton()
    btHook.setText('Hook')
    btHook.clicked.connect(self.hook)
    self.lbHook = lbHook = QLabel()
    lbHook.setText('')
    llHook = HBox(
      btHook,
      lbHook,
      None,
    )
    ## tab
    qtab = self.qtab = QTabWidget(self)
    self.subwWFC = WFCWidget(dolphin=self.dolphin)
    self.tidxWFC = qtab.addTab(self.subwWFC, 'WFC')
    self.subwRuntime = RuntimeWidget(dolphin=self.dolphin)
    self.tidxRuntime = qtab.addTab(self.subwRuntime, 'Runtime')
    qtab.currentChanged.connect(lambda: self.onTabChanged(qtab.currentIndex()))
    # dummy widget to put widgets in the center
    w = QWidget()
    w.setLayout(VBox(
      llHook,
      qtab,
      None,
    ))
    self.setCentralWidget(w)
  def hook(self):
    err = self.dolphin.hook()
    self.lbHook.setText('Hooked' if err is None else err)
    self.qtab.setEnabled(self.dolphin.hooked)
  def onTabChanged(self, index):
    self.subwWFC.activate(index==self.tidxWFC)
